app/uk/gov/hmrc/economiccrimelevyreturns/connectors/BaseConnector.scala

```diff
@@ -18,23 +18,66 @@ package uk.gov.hmrc.economiccrimelevyreturns.connectors
 
 import play.api.http.Status.{ACCEPTED, CREATED, OK}
 import play.api.libs.json.{JsResult, Reads}
+import uk.gov.hmrc.http.client.RequestBuilder
 import uk.gov.hmrc.http.{HttpResponse, UpstreamErrorResponse}
 
 import scala.concurrent.{ExecutionContext, Future}
 
 trait BaseConnector {
-  def executeAndDeserialise[T](
-    response: Future[HttpResponse]
-  )(implicit ec: ExecutionContext, reads: Reads[T]): Future[T] = response.flatMap { response =>
-    response.status match {
-      case OK | CREATED | ACCEPTED =>
-        response.json
-          .validate[T]
-          .map(result => Future.successful(result))
-          .recoverTotal(error => Future.failed(JsResult.Exception(error)))
-      case _                       =>
-        Future.failed(UpstreamErrorResponse(response.body, response.status))
-    }
+
+  def retryCondition: PartialFunction[Exception, Boolean] = {
+    case e: UpstreamErrorResponse if UpstreamErrorResponse.Upstream5xxResponse.unapply(e).isDefined => true
   }
 
+  implicit class HttpResponseHelpers(response: HttpResponse) {
+
+    def error[A]: Future[A] =
+      Future.failed(UpstreamErrorResponse(response.body, response.status))
+
+    def as[A](implicit reads: Reads[A]): Future[A] =
+      response.json
+        .validate[A]
+        .map(result => Future.successful(result))
+        .recoverTotal(error => Future.failed(JsResult.Exception(error)))
+
+    def asOption[A](implicit reads: Reads[A]): Future[Option[A]] =
+      response.json
+        .validateOpt[A]
+        .map(result => Future.successful(result))
+        .recoverTotal(error => Future.failed(JsResult.Exception(error)))
+  }
+
+  implicit class RequestBuilderHelpers(requestBuilder: RequestBuilder) {
+    def executeAndDeserialise[T](implicit ec: ExecutionContext, reads: Reads[T]): Future[T] =
+      requestBuilder
+        .execute[HttpResponse]
+        .flatMap { response =>
+          response.status match {
+            case OK | CREATED | ACCEPTED => response.as[T]
+            case _                       =>
+              response.error
+          }
+        }
+
+    def executeAndDeserialiseOpt[T](implicit ec: ExecutionContext, reads: Reads[T]): Future[Option[T]] =
```

Comment:
Can we use `OptionT[Future, T]` instead of `Future[Option[T]]` ?

Reply:
Hell yeah!